README.rst
===============================================================================
Socrates
===============================================================================

Socrates is a simple static site generator. It's geared towards blogs. You
write your posts in your favorite plain text to HTML language (e.g. Markdown,
textile, reStructuredText) and save them as text files on your harddrive.
Socrates then takes them, and creates a full HTML site for you. For free, you
will get a home page which lists latest posts, single post pages, category
pages, archive pages, an about page and an atom feed.

Features
-------------------------------------------------------------------------------

* Familiar Django and Jinja2 templates
* Simple install via pip
* Markdown, reStructuredText, Textile support
* YAML configuration
* Atom feed
* Github pages compatible
* Real HTML punctuation

Usage
-------------------------------------------------------------------------------

First, you need to create a new blog::

    $ socrates -i blog

This will create a ``blog`` directory with a simple blog structure::

    blog
        posts
            2010-your-post.md
        pages
            about.md
            contact.md
        layout
            index.html
            single.html
            category.html
            ...
        media
            style.css
        config.yaml

The ``posts`` directory is where you will place your posts files. Anything
prefixed with ``_`` or ``.`` will be ignored. ``layout`` is your basic theme or
a template. ``config.yaml`` is a site-wide configuration file. Don't forget to
update the about file with relevant information.

You can also create a new blog in the current working directory::

    $ socrates -i

When you are ready to generate your site, you run::

    $ socrates -g blog

Or, ::

    $ socrates -g

for current directory.

This will place all the generated files in ``blog/deploy``. You can then take
that directory and upload it to your server.

Installation
-------------------------------------------------------------------------------

First, set up your virtual environment and pip install Socrates.

::

    virtualenv env --no-site-packages
    source env/bin/activate
    pip install socrates

Or, to get the latest code, you can install from Github or Bitbucket.

::

    pip install -e git://github.com/honza/socrates.git#egg=socrates

The next step is to install your templates. At the moment, you can choose
between `Django templates`_ and `Jinja2 templates`_. The default theme uses
Django.

Themes
-------------------------------------------------------------------------------

Socrates supports Django templates and Jinja2 templates. You can specify which
templating engine you wish to use in the ``config.yaml`` file. There are two
basic themes in the ``themes`` directory to get you started.

Text to HTML
-------------------------------------------------------------------------------

Socrates will attemp to convert your text files into HTML using a text
processor. You should specify the name of the text processor in
``config.yaml``. It should be all lowercase. Currently supported processors:

  - Markdown
  - Textile
  - reStructuredText
  - HTML (unmodified text)

Development
-------------------------------------------------------------------------------

You can start a simple development server to aid you in development.::

    $ socrates -r blog

Or, ::

    $ socrates -r

for current directory.

During development and/or composing a layout theme, it might be useful to always re-generate all posts, without caching.
To do so, you can run the following option: ::

    $ socrates -c


Documentation
-------------------------------------------------------------------------------

The documentation is contained within the ``docs`` directory and is written in
reStructuredText using `Sphinx`_. The documentation is easily read in a
standard text editor. However, you can build an HTML version like so::

    $ pip install sphinx
    $ cd docs/
    $ make html
    $ open _build/html/index.html

Or, you can view the online version of the latest `documentation`_.


License
-------------------------------------------------------------------------------

Socrates is licensed under the terms of the 3-clause BSD license.

Contribute
-------------------------------------------------------------------------------

All contributions are welcome. Please do ask if a feature is wanted before
starting work to avoid wasting your time.

Bugs & Issues
-------------------------------------------------------------------------------

Please report all bugs on Github.  `Mailing list`_.  You can also come visit us
in ``#socrates`` on freenode.

Authors
-------------------------------------------------------------------------------

Socrates was originally written by Honza Pokorny. For a complete list of
authors, please see the AUTHORS file.

In the wild
-------------------------------------------------------------------------------

* `Honza's blog <http://honza.ca>`_
* `Adam's blog <http://thurloat.com>`_

.. _Django templates: https://docs.djangoproject.com/en/1.3/#the-template-layer
.. _Jinja2 templates: http://jinja.pocoo.org/docs/
.. _Sphinx: http://sphinx.pocoo.org/
.. _documentation: http://readthedocs.org/docs/socrates/en/latest/index.html
.. _Mailing list: https://groups.google.com/group/python-socrates


CHANGELOG
==== 0.9.2 (2016-03-01) ====
	- Better parsing errors

==== 0.9.1 (2015-11-12) ====
	- Misaka 2.0
	- cache control

==== 0.9.0 (2013-07-26) ====
	- Python 3 support
	- Atom feed fixes
	- Add in_category template
	- Cache posts and rebuild when needed
	- Fix encoding issues (thurloat)
	- Allow user to specify deploy dir
	- Fix misaka support

==== 0.8.0 (2012-02-12) ====
	- Add silent flag
	- Write errors to stderr
	- Add punctuation via smartypants and typogrify
	- Add page processing
	- Update theme layout

==== 0.7.0 (2012-01-24) ====
	- Better exception handling
	- Support for pygments in Markdown
	- Use misaka for markdown
	- Add sitemap logic (thurloat)
	- Custom slugs (thurloat)
	- raw html renderer (thurloat)

==== 0.6.4 (2011-09-07) ====
	- Packaging

==== 0.6.3 (2011-09-07) ====
	- Packaging

==== 0.6.2 (2011-09-07) ====
	- Hot fix file_type errors

==== 0.6.1 (2011-07-25) ====
	- Hot fix (unable to generate rst-based posts)

==== 0.6.0 (2011-07-23) ====
	- Don't add pygments styles inline
	- Extension based processing
	- Allow skipping of archives and categories
	- Register Socrates in Pypi
	- Add pygments settings (thurloat)
	- Fix dev server (thurloat)
	- Mark posts with no categories as 'Uncategorized' (thurloat)

==== 0.5 (2011-07-20) ====
	- Add append slash setting
	- Github pages support
	- rst native support (with pygments)
	- Per post template overrides
	- Remove many dependencies
	- Command-line usage simplification
	- Allow day in urls (thurloat)
	- Make dev server serve files without .html extension (thurloat)
	- Unicode fixes (thurloat)

==== 0.4 (2011-07-18) ====
	- pip install support
	- socrates executable

==== 0.3 (2011-05-27) ====
	- add jinja2 support
	- add new theme
	- about page support
	- atom feed support
	- Use BSD instead of GPL

==== 0.2 (2011-02-26) ====
	- complete rewrite
	- faster execution
	- socrates is now only one file
	- add markdown, textile and reStructuredText
	- remove pygments as dependency

==== 0.1 (2011-02-25) ====
	- initial release


socrates/socrates.py
"""
Socrates - Static site generator
(c) 2011-2015 - Honza Pokorny et al.
http://github.com/honza/socrates
BSD licensed
"""
import os
import sys
import shutil
import yaml
import json
from datetime import datetime

from .renderers import DjangoRenderer, Jinja2Renderer
from .models import Post, Page
from .utils import slugify
from .exceptions import ConfigurationError


DEFAULTS = {
    'author': 'author',
    'site_name': 'Socrates site',
    'posts_per_page': 10,
    'url': 'http://example.com',
    'date_format': '%B %d, %Y',
    'text_processor': 'markdown',
    'templates': 'django',
    'append_slash': True,
    'url_include_day': False,
    'initial_header_level': 2,
    'skip_archives': False,
    'skip_categories': False,
    'inline_css': False,
    'pygments': {},
    'sitemap_urls': ['about'],
    'punctuation': False,
    'ligatures': False,
    'deploy_dir': 'deploy'
}


AVAILABLE_EXTENSIONS = ['markdown', 'rst', 'textile', 'html', 'extension']
POST_CACHE_FILENAME = '.post-cache.json'


class Generator(object):
    """
    This is the main class of the application. It handles the collection,
    processing and rendering of the site.
    """

    # Templates
    SINGLE = 'single.html'
    PAGE = 'page.html'
    INDEX = 'index.html'
    CATEGORY = 'category.html'
    ARCHIVE = 'archive.html'
    PAGED = 'index_paged.html'
    ATOM = 'atom.html'
    SITEMAP = 'sitemap.xml'

    # Directories
    ROOT = None
    DEPLOY = None
    POSTS = None
    PAGES = None

    # Global, site-wide settings
    SETTINGS = None

    def __init__(self, directory, silent=False):
        m = os.getcwd()
        self.ROOT = os.path.join(m, directory)
        self.silent = silent

        if not os.path.exists(self.ROOT):
            sys.stderr.write("The '%s' directory doesn't exist.\n" % directory)
            return
        self.SETTINGS = self._get_settings()
        # Set up deploy directory
        self.DEPLOY = os.path.join(self.ROOT, self.SETTINGS['deploy_dir'])
        self.deploy_dir_created = False
        if not os.path.exists(self.DEPLOY):
            os.mkdir(self.DEPLOY)
            self.deploy_dir_created = True
        # Copy media files to deploy destination
        if os.path.exists(os.path.join(self.DEPLOY, 'media')):
            shutil.rmtree(os.path.join(self.DEPLOY, 'media'))
        shutil.copytree(os.path.join(self.ROOT, 'layout', 'media'),
                        os.path.join(self.DEPLOY, 'media'))

        self.POSTS = os.path.join(self.ROOT, 'posts')
        self.PAGES = os.path.join(self.ROOT, 'pages')

        if self.SETTINGS['text_processor'] not in AVAILABLE_EXTENSIONS:
            ext = self.SETTINGS['text_processor']
            sys.stderr("WARNING: %s isn't a recognizeed text processor.\n"
                       % ext)
            sys.exit(1)
        self.init_template_renderer()

        self.posts = []
        self.pages = []
        self.categories = {}
        self.years = {}
        self.archives = {}

        self.load_posts()
        self.cache_data = self.get_post_cache()
        self.process_posts()

        if os.path.exists(self.PAGES):
            self.load_pages()
            self.process_pages()

        self.SETTINGS = dict(self.SETTINGS, **{'categories': self.categories})
        self.SETTINGS = dict(self.SETTINGS, **{'years': self.years})

        # Create html files
        self.make_index_page()
        self.make_atom()
        self.make_sitemap()

        if not self.SETTINGS['skip_categories']:
            self.make_category_pages()
        if not self.SETTINGS['skip_archives']:
            self.make_archive_pages()

        self.make_pagination()

        if not self.SETTINGS['inline_css']:
            from pygments.formatters import HtmlFormatter
            formatter = HtmlFormatter(**self.SETTINGS['pygments'])
            css_content = formatter.get_style_defs('.highlight')
            self._write_to_file(
                os.path.join(self.DEPLOY, 'media', 'pygments.css'),
                css_content)

        self.save_post_cache()
        self.log("Success!")

    def init_template_renderer(self):
        """
        Create and save a reference to an instance of a template renderer
        """
        t = self.SETTINGS['templates']
        layout = os.path.join(self.ROOT, 'layout')
        if t == 'jinja2':
            self.template = Jinja2Renderer(layout)
        elif t == 'django':
            self.template = DjangoRenderer(layout)
        else:
            raise NotImplementedError("I don't know this template type.")

    def render(self, template, values):
        """
        Give a template name and a dict of values, return an HTML
        representation of a page/post.
        """
        return self.template.render(template, values)

    def log(self, text):
        """
        A thin wrapper around the ``print`` statement to allow silencing of all
        non-error output.
        """
        if not self.silent:
            print(text)

    def _get_page_str(self, current, total):
        """
        Return page number as a string for pagination.
        """
        current += 1
        if total < 10:
            return str(current)
        elif 10 <= total < 100:
            if total < 10:
                return '0%d' % current
            else:
                return str(current)

    def _get_post_cats(self, post):
        """
        Extract and save post's categories
        """
        cats = post.config.get('categories', ['Uncategorized'])
        for c in cats:
            if c not in self.categories:
                self.categories[c] = [post]
            else:
                self.categories[c].append(post)

    def _get_settings(self):
        """
        Read main config file. Return dict.
        """
        s = os.path.join(self.ROOT, 'config.yaml')
        if not os.path.exists(s):
            raise Exception('No config file.')
        c = open(s, 'r').read()
        return dict(DEFAULTS, **yaml.load(c))

    def _write_to_file(self, path, contents):
        """
        Create a file (path) with contents
        """
        f = open(path, 'w')
        f.write(contents.encode("utf-8"))
        f.close()

    def _v(self, vals):
        """
        Create a joint dict of global config and local template vals
        """
        return dict(self.SETTINGS, **vals)

    def load_posts(self):
        """
        Get all files from the posts directory, create Post instances and add
        them to the self.posts list.
        """
        for filename in os.listdir(self.POSTS):
            if not filename.startswith('.') and not filename.startswith('_'):
                p = os.path.join(self.POSTS, filename)
                try:
                    self.posts.append(Post(p, self.SETTINGS))
                except ConfigurationError:
                    sys.stderr.write("WARNING: %s isn't configured properly.\n"
                                     % filename)
                    sys.exit(1)
                except:
                    sys.stderr.write("ERROR: couldn't parse %s.\n"
                                     % filename)
                    sys.exit(1)

        self.posts.reverse()

    def load_pages(self):
        """
        Get all files from the posts directory, create Post instances and add
        them to the self.posts list.
        """
        for filename in os.listdir(self.PAGES):
            if not filename.startswith('.') and not filename.startswith('_'):
                p = os.path.join(self.PAGES, filename)
                try:
                    self.pages.append(Page(p, self.SETTINGS))
                except ConfigurationError:
                    sys.stderr.write("WARNING: %s isn't configured properly.\n"
                                     % filename)
                    sys.exit(1)

    def process_posts(self):
        """
        Collect all the necessary information about posts.
        """
        self.log('Processing posts...')
        for post in self.posts:

            # Get categories
            self._get_post_cats(post)

            # Post dirs
            #   years
            date = post.config['date']
            year = str(date.year)
            if year not in self.years:
                self.years[year] = {}
            #   months
            m = date.strftime("%m")
            y = self.years[year]
            if m not in list(y.keys()):
                self.years[year][m] = []

            if self.SETTINGS['url_include_day']:
                if post.day not in self.years[year][m]:
                    self.years[year][m].append(post.day)

            # Archives
            d = year
            if d not in self.archives:
                self.archives[d] = [post]
            else:
                self.archives[d].append(post)

        self.make_post_directories()
        self.save_posts()

    def process_pages(self):
        self.save_pages()

    def save_posts(self):
        self.log('Saving posts...')
        for post in self.posts:

            # TODO: All of this caching needs to happen much sooner than here
            # because the posts have already been parsed.
            try:
                old_hash = self.cache_data[post.path]
            except KeyError:
                old_hash = None

            new_hash = post.hash_file()
            self.cache_data[post.path] = new_hash

            if old_hash == new_hash:
                continue

            # Save the thing
            b = post.slug + '.html'

            if self.SETTINGS['url_include_day']:
                m = os.path.join(self.DEPLOY, post.year, post.month,
                                 post.day, b)
            else:
                m = os.path.join(self.DEPLOY, post.year, post.month, b)

            if 'template' in list(post.config.keys()):
                t = post.config['template']
            else:
                t = self.SINGLE

            content = self.render(t, self._v({'post': post}))

            # Print filename to show progress
            self.log(post.filename)

            self._write_to_file(m, content)

    def save_pages(self):
        self.log('Saving pages...')
        for page in self.pages:
            # Save the thing
            b = page.slug + '.html'

            m = os.path.join(self.DEPLOY, b)

            if 'template' in list(page.config.keys()):
                t = page.config['template']
            else:
                t = self.PAGE

            content = self.render(t, self._v({'page': page}))

            # Print filename to show progress
            self.log(page.filename)
            self._write_to_file(m, content)

    def make_post_directories(self):
        """
        If they don't exist, create post directories.
            2011
                01
                  04
                  05
                  12
                03
                  09
                  11
                12
                  21
            2010
                04
                  11
                07
                  30
        """
        self.log('Creating directories...')
        keys = list(self.years.keys())
        for k in keys:
            m = os.path.join(self.DEPLOY, k)
            if not os.path.exists(m):
                os.mkdir(m)

            months = self.years[k]
            for month in months:
                m = os.path.join(self.DEPLOY, k, month)
                if not os.path.exists(m):
                    os.mkdir(m)

                if self.SETTINGS['url_include_day']:
                    days = self.years[k][month]
                    for day in days:
                        m = os.path.join(self.DEPLOY, k, month, day)
                        if not os.path.exists(m):
                            os.mkdir(m)

    def make_index_page(self):
        """
        Create an index page
        """
        m = os.path.join(self.DEPLOY, 'index.html')
        if self.SETTINGS['posts_per_page'] == 0:
            # List all posts on the index page
            posts = self.posts
            extra = False
        elif len(self.posts) > self.SETTINGS['posts_per_page']:
            # Cut off unnecessary posts
            n = len(self.posts)
            n = self.SETTINGS['posts_per_page']
            posts = self.posts[:n]
            extra = True
        else:
            posts = self.posts
            extra = False
        contents = self.render(self.INDEX,
                               self._v({'posts': posts, 'extra': extra}))
        self._write_to_file(m, contents)

    def make_sitemap(self):
        """
        Generate a sitemap.
        """
        m = os.path.join(self.DEPLOY, 'sitemap.xml')
        posts = self.posts
        contents = self.render(self.SITEMAP, self._v({
            'posts': posts,
            'now': self._get_atom_date()}))
        self._write_to_file(m, contents)

    def make_atom(self):
        """
        Create an atom feed
        """
        m = os.path.join(self.DEPLOY, 'atom.xml')
        if self.SETTINGS['posts_per_page'] == 0:
            # List all posts on the index page
            posts = self.posts
        elif len(self.posts) > self.SETTINGS['posts_per_page']:
            # Cut off unnecessary posts
            n = len(self.posts)
            n = self.SETTINGS['posts_per_page']
            posts = self.posts[:n]
        else:
            posts = self.posts
        contents = self.render(self.ATOM,
                               self._v({'posts': posts,
                                        'now': self._get_atom_date()}))

        contents = contents.replace('&nbsp;', '')
        self._write_to_file(m, contents)

    def _get_atom_date(self):
        date = datetime.utcnow()
        d = date.strftime('%Y-%m-%dT%H:%M:%S%z')
        return d + "Z"

    def make_category_pages(self):
        """
        Make category pages. They go into the 'category' directory.
        """
        keys = list(self.categories.keys())
        if len(keys) != 0:
            # Make category dir
            self.CATEGORIES = os.path.join(self.DEPLOY, 'category')
            if not os.path.exists(self.CATEGORIES):
                os.mkdir(self.CATEGORIES)
            for k in keys:
                if not k:
                    continue
                p = os.path.join(self.CATEGORIES, slugify(k))
                if not os.path.exists(p):
                    os.mkdir(p)
                posts = self.categories[k]
                contents = self.render(
                    self.CATEGORY, self._v({'category': k, 'posts': posts}))
                m = os.path.join(p, 'index.html')
                self._write_to_file(m, contents)

    def make_archive_pages(self):
        """
        Make archive pages. Only by year.
        """
        self.log('Creating archives...')
        keys = list(self.archives.keys())
        if len(keys) != 0:
            # Make category dir
            self.ARCHIVES = os.path.join(self.DEPLOY, 'archive')
            if not os.path.exists(self.ARCHIVES):
                os.mkdir(self.ARCHIVES)
            for k in keys:
                p = os.path.join(self.ARCHIVES, slugify(k))
                if not os.path.exists(p):
                    os.mkdir(p)
                posts = self.archives[k]
                contents = self.render(
                    self.ARCHIVE, self._v({'year': k, 'posts': posts}))
                m = os.path.join(p, 'index.html')
                self._write_to_file(m, contents)

    def make_pagination(self):
        """
        Make pagination.
            /page/2/
            /page/3/
        Uses posts per page setting
        """
        per = int(self.SETTINGS['posts_per_page'])
        if per == 0:
            # Skip pagination if all posts are on index page
            return
        self.log('Creating pagination...')
        num = len(self.posts)
        pages = num / per

        m = os.path.join(self.DEPLOY, 'page')
        if not os.path.exists(m):
            os.mkdir(m)
        for x in range(1, pages + 1):
            if x == 1:
                prev = '/'
            else:
                prev = '/page/%d/' % x

            if x < pages:
                next = '/page/%d/' % int(x + 1)
            else:
                next = None

            n = per * x
            p = self.posts[n:n + per]

            v = {
                'page': x + 1,
                'posts': p,
                'prev': prev,
                'total': pages + 1,
                'next': next
            }

            folder = self._get_page_str(x, pages + 1)

            e = os.path.join(m, folder)
            if not os.path.exists(e):
                os.mkdir(e)

            contents = self.render(self.PAGED, self._v(v))
            c = os.path.join(e, "index.html")
            self._write_to_file(c, contents)

    def get_post_cache(self):
        if not os.path.exists(POST_CACHE_FILENAME) or self.deploy_dir_created:
            return {}

        return json.loads(open(POST_CACHE_FILENAME).read())

    def save_post_cache(self):
        self.log("Saving cache")
        with open(POST_CACHE_FILENAME, 'w') as f:
            f.write(json.dumps(self.cache_data, indent=4))
